Update evaluation docs (#11173)
PyTorchLightning/pytorch-lightning: docs/source/common/test_set.rst

.. _test_set:

Test set

> ========

Reviewer: This page is a bit confusing. We should rename it `evaluation.rst`.

Author: yes. forgot to mention. will be doing that once approved for better review now.

Reviewer 2: won't that break existing links?

Author: yes.. not updating now. will be renaming them at the very end once all docs are done.

##########

> *******

Reviewer: I think we should start by best practices around validation vs test and then introduce how to use them to perform evaluation.

Testing
*******

Lightning allows the user to test their models with any compatible test dataloaders. This can be done before/after training
and is completely agnostic to :meth:`~pytorch_lightning.trainer.trainer.Trainer.fit` call. The logic used here is defined under
:meth:`~pytorch_lightning.core.lightning.LightningModule.validation_step`.

Testing is performed using the ``Trainer`` object's ``.test()`` method.

.. automethod:: pytorch_lightning.trainer.Trainer.test
    :noindex:

> Test after Fit

Reviewer: We should add a note that benchmarking for papers with the test should be done with only a single device due to the DistributedSampler.

Author: good point! will mention

.. _test_set:

##########
Evaluation
##########

During and after training we need a way to evaluate our models to make sure they are not overfitting while training and
generalize well on unseen or real-world data. There are generally 2 stages of evaluation: validation and testing. To some
degree they serve the same purpose, to make sure models works on real data but they have some practical differences.

Validation is usually done during training, traditionally after each training epoch. It can be used for hyperparameter optimization or tracking model performance during training.
It's a part of the training process.

> Testing is usually done once we are satisfied with the training and use the best model selected from the validation metrics and want to evaluate its performance on unseen data where the model acts
> as a black box. The test dataset is emulating the production/benchmark data where labels aren't available.

Reviewer 2: ```suggestion
Testing is usually done once we are satisfied with the training and only with the best model selected from the validation metrics.
```

There's a lot of nuance here in reality so I would keep these explanations as short as possible. We want to motivate the functionality, not teach ML here

In this  case, the options you pass to trainer will be used when
running the test set (ie: 16-bit, dp, ddp, etc...)

Test with Additional DataLoaders
================================

> You can still run inference on a test set even if the :meth:`~pytorch_lightning.core.hooks.DataHooks.test_dataloader` method hasn't been

Reviewer: ```suggestion
You can still run inference on a test dataset even if the :meth:`~pytorch_lightning.core.hooks.DataHooks.test_dataloader` method hasn't been
```

    # test (pass in datamodule)
    trainer.test(datamodule=dm)

----------

**********
Validation
**********

> Lightning allows the user to validate their models with any compatible val dataloaders. This can be done before/after training.

Reviewer: ```suggestion
Lightning allows the user to validate their models with any compatible ``val dataloaders``. This can be done before/after training.
```

> The logic used here is defined under :meth:`~pytorch_lightning.core.lightning.LightningModule.validation_step`.

Reviewer: ```suggestion
The logic associated to the validation is defined within the :meth:`~pytorch_lightning.core.lightning.LightningModule.validation_step`.
```

> Apart from this ``.validate`` has same API as that of ``.test`` as defined above.

Reviewer: ```suggestion
Apart from this ``.validate`` has same API as ``.test``, but would rely respectively on :meth:`~pytorch_lightning.core.lightning.LightningModule.validation_step` and :meth:`~pytorch_lightning.core.lightning.LightningModule.test_step`.
```

> .. note::

Author: not required. it renders fine without an extra line.
